// node-zoho/build/products/crm/custommodule1.js
// Generated by CoffeeScript 2.2.1
var CrmModule, CustomModule1;

CrmModule = require('./crm-module');

CustomModule1 = (function() {
  class CustomModule1 extends CrmModule {
    getMyRecords() {
      throw new Error('Not Implemented');
    }

    getSearchRecordsByPDC() {
      throw new Error('Not Implemented');
    }

    deleteRecords() {
      throw new Error('Not Implemented');
    }

    convertLead() {
      throw new Error('Not Implemented');
    }

    getRelatedRecords() {
      throw new Error('Not Implemented');
    }

    updateRelatedRecords() {
      throw new Error('Not Implemented');
    }

    getUsers() {
      throw new Error('Not Implemented');
    }

    downloadFile() {
      throw new Error('Not Implemented');
    }

    deleteFile() {
      throw new Error('Not Implemented');
    }

    uploadPhoto() {
      throw new Error('Not Implemented');
    }

    downloadPhoto() {
      throw new Error('Not Implemented');
    }

    deletePhoto() {
      throw new Error('Not Implemented');
    }

  };

  CustomModule1.prototype.name = 'CustomModule1';

  return CustomModule1;

}).call(this);

module.exports = CustomModule1;
