// node-zoho/build/products/crm/index.js
// Generated by CoffeeScript 2.2.1
var BaseProduct, CrmProduct;

BaseProduct = require('../../base-product');

CrmProduct = (function() {
  class CrmProduct extends BaseProduct {
    getModules() {
      return ['Leads', 'Events', 'Potentials', 'Contacts', 'Accounts', 'CustomModule1'];
    }

    getModule(module_name) {
      var err, instance, module_class, module_path;
      try {
        module_name = module_name.toLowerCase();
        module_path = `./${module_name}`;
        require.resolve(module_path);
        module_class = require(module_path);
      } catch (error) {
        err = error;
        throw new Error(err);
      }
      instance = new module_class(this);
      return instance;
    }

    getScope() {
      return 'crmapi';
    }

    getBaseUrl() {
      return {
        hostname: `crm.zoho.${this.zoho.region}`,
        protocol: 'https',
        query: {
          authtoken: this.zoho.authToken,
          scope: this.getScope()
        },
        path: [this.name]
      };
    }

  };

  CrmProduct.prototype.name = 'crm';

  return CrmProduct;

}).call(this);

module.exports = CrmProduct;
