// node-zoho/build/base-module.js
// Generated by CoffeeScript 2.2.1
var BaseModule, _;

_ = require('underscore');

BaseModule = (function() {
  class BaseModule {
    constructor(product) {
      this.product = product;
      if (!this.product) {
        throw new Error('Requires product');
      }
      return;
    }

    getUrlParts() {
      return [this.scope, this.format, this.name];
    }

    buildUrl(query = {}, path = [], options = {}) {
      var method, url;
      method = 'POST';
      if (options != null ? options.method : void 0) {
        method = options.method.toUpperCase();
      }
      url = this.product.getBaseUrl();
      url.method = method;
      url.path = url.path.concat(this.getUrlParts());
      url.path = url.path.concat(path);
      url.pathname = "/" + url.path.join("/");
      _.extend(url.query, query);
      return url;
    }

  };

  BaseModule.prototype.name = "BaseModule";

  BaseModule.prototype.format = "xml";

  BaseModule.prototype.scope = "private";

  return BaseModule;

}).call(this);

module.exports = BaseModule;
