// node-zoho/build/node-zoho.js
// Generated by CoffeeScript 2.2.1
var Zoho;

Zoho = (function() {
  class Zoho {
    constructor(options = {}) {
      if (options.region != null) {
        this.region = options.region;
      }
      this.authDefaults = {
        host: `accounts.zoho.${this.region}`,
        port: 443,
        path: "/apiauthtoken/nb/create?SCOPE=ZohoCRM/crmapi"
      };
      if (options != null ? options.authToken : void 0) {
        this.authToken = options != null ? options.authToken : void 0;
      }
      return this;
    }

    getProduct(productName) {
      var Crm;
      Crm = require('./products/' + productName);
      return new Crm(this);
    }

    execute(product, module, call) {
      var args, moduleInstance, productInstance;
      args = Array.prototype.slice.call(arguments);
      productInstance = this.getProduct(product);
      moduleInstance = productInstance.getModule(module);
      return moduleInstance[call].apply(moduleInstance, args.slice(3));
    }

  };

  Zoho.prototype.authToken = null;

  Zoho.prototype.region = 'com';

  return Zoho;

}).call(this);

module.exports = Zoho;
